[X86] Try Folding icmp of v8i32 -> fcmp of v8f32 on AVX

goldsteinn opened the pull request:
Fixes: #82242

The idea is that AVX doesn't support comparisons for `v8i32` so it
splits the comparison into 2x `v4i32` comparisons + reconstruction of
the `v8i32`.

By converting to a float, we can handle the comparison with 1/2
instructions (1 if we can `bitcast`, 2 if we need to cast with
`sitofp`).

The Proofs: https://alive2.llvm.org/ce/z/AJDdQ8
Timeout, but they can be reproduced locally.

goldsteinn committed:
[X86] Add tests for folding `icmp` of `v8i32` -> `fcmp` of `v8f32` on AVX; NFC

goldsteinn commented:
NB: Proofs of the transform: https://alive2.llvm.org/ce/z/AJDdQ8

They timeout online, so need to be run locally. here is the output:
```
; $> /home/noah/programs/opensource/llvm-dev/src/alive2/build/alive-tv (-smt-to=200000000)

----------------------------------------
define i1 @src_eq(i32 %x, i32 %C) {
#0:
  %C_abs = abs i32 %C, 0
  %X_abs = abs i32 %x, 0
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp eq i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_eq(i32 %x, i32 %C) {
#0:
  %C_abs = abs i32 %C, 0
  %X_abs = abs i32 %x, 0
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = sitofp i32 %C to float
  %conv = sitofp i32 %x to float
  %cmp = fcmp oeq float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_ne(i32 %x, i32 %C) {
#0:
  %C_abs = abs i32 %C, 0
  %X_abs = abs i32 %x, 0
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp ne i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_ne(i32 %x, i32 %C) {
#0:
  %C_abs = abs i32 %C, 0
  %X_abs = abs i32 %x, 0
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = sitofp i32 %C to float
  %conv = sitofp i32 %x to float
  %cmp = fcmp one float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_slt(i32 %x, i32 %C) {
#0:
  %C_abs = abs i32 %C, 0
  %X_abs = abs i32 %x, 0
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp slt i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_slt(i32 %x, i32 %C) {
#0:
  %C_abs = abs i32 %C, 0
  %X_abs = abs i32 %x, 0
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = sitofp i32 %C to float
  %conv = sitofp i32 %x to float
  %cmp = fcmp olt float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_sgt(i32 %x, i32 %C) {
#0:
  %C_abs = abs i32 %C, 0
  %X_abs = abs i32 %x, 0
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp sgt i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_sgt(i32 %x, i32 %C) {
#0:
  %C_abs = abs i32 %C, 0
  %X_abs = abs i32 %x, 0
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = sitofp i32 %C to float
  %conv = sitofp i32 %x to float
  %cmp = fcmp ogt float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_sle(i32 %x, i32 %C) {
#0:
  %C_abs = abs i32 %C, 0
  %X_abs = abs i32 %x, 0
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp sle i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_sle(i32 %x, i32 %C) {
#0:
  %C_abs = abs i32 %C, 0
  %X_abs = abs i32 %x, 0
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = sitofp i32 %C to float
  %conv = sitofp i32 %x to float
  %cmp = fcmp ole float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_sge(i32 %x, i32 %C) {
#0:
  %C_abs = abs i32 %C, 0
  %X_abs = abs i32 %x, 0
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp sge i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_sge(i32 %x, i32 %C) {
#0:
  %C_abs = abs i32 %C, 0
  %X_abs = abs i32 %x, 0
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = sitofp i32 %C to float
  %conv = sitofp i32 %x to float
  %cmp = fcmp oge float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_ult(i32 %x, i32 %C) {
#0:
  %C_lemma0 = icmp ult i32 %C, 16777216
  %X_lemma0 = icmp ult i32 %x, 16777216
  %lemma0 = or i1 %C_lemma0, %X_lemma0
  assume i1 %lemma0
  %C_lemma1 = icmp sge i32 %C, 0
  %X_lemma1 = icmp sge i32 %x, 0
  %lemma1 = and i1 %C_lemma1, %X_lemma1
  assume i1 %lemma1
  %cmp = icmp ult i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_ult(i32 %x, i32 %C) {
#0:
  %C_lemma0 = icmp ult i32 %C, 16777216
  %X_lemma0 = icmp ult i32 %x, 16777216
  %lemma0 = or i1 %C_lemma0, %X_lemma0
  assume i1 %lemma0
  %C_lemma1 = icmp sge i32 %C, 0
  %X_lemma1 = icmp sge i32 %x, 0
  %lemma1 = and i1 %C_lemma1, %X_lemma1
  assume i1 %lemma1
  %CFp = sitofp i32 %C to float
  %conv = sitofp i32 %x to float
  %cmp = fcmp olt float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_ule(i32 %x, i32 %C) {
#0:
  %C_lemma0 = icmp ult i32 %C, 16777216
  %X_lemma0 = icmp ult i32 %x, 16777216
  %lemma0 = or i1 %C_lemma0, %X_lemma0
  assume i1 %lemma0
  %C_lemma1 = icmp sge i32 %C, 0
  %X_lemma1 = icmp sge i32 %x, 0
  %lemma1 = and i1 %C_lemma1, %X_lemma1
  assume i1 %lemma1
  %cmp = icmp ule i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_ule(i32 %x, i32 %C) {
#0:
  %C_lemma0 = icmp ult i32 %C, 16777216
  %X_lemma0 = icmp ult i32 %x, 16777216
  %lemma0 = or i1 %C_lemma0, %X_lemma0
  assume i1 %lemma0
  %C_lemma1 = icmp sge i32 %C, 0
  %X_lemma1 = icmp sge i32 %x, 0
  %lemma1 = and i1 %C_lemma1, %X_lemma1
  assume i1 %lemma1
  %CFp = sitofp i32 %C to float
  %conv = sitofp i32 %x to float
  %cmp = fcmp ole float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_ugt(i32 %x, i32 %C) {
#0:
  %C_lemma0 = icmp ult i32 %C, 16777216
  %X_lemma0 = icmp ult i32 %x, 16777216
  %lemma0 = or i1 %C_lemma0, %X_lemma0
  assume i1 %lemma0
  %C_lemma1 = icmp sge i32 %C, 0
  %X_lemma1 = icmp sge i32 %x, 0
  %lemma1 = and i1 %C_lemma1, %X_lemma1
  assume i1 %lemma1
  %cmp = icmp ugt i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_ugt(i32 %x, i32 %C) {
#0:
  %C_lemma0 = icmp ult i32 %C, 16777216
  %X_lemma0 = icmp ult i32 %x, 16777216
  %lemma0 = or i1 %C_lemma0, %X_lemma0
  assume i1 %lemma0
  %C_lemma1 = icmp sge i32 %C, 0
  %X_lemma1 = icmp sge i32 %x, 0
  %lemma1 = and i1 %C_lemma1, %X_lemma1
  assume i1 %lemma1
  %CFp = sitofp i32 %C to float
  %conv = sitofp i32 %x to float
  %cmp = fcmp ogt float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_uge(i32 %x, i32 %C) {
#0:
  %C_lemma0 = icmp ult i32 %C, 16777216
  %X_lemma0 = icmp ult i32 %x, 16777216
  %lemma0 = or i1 %C_lemma0, %X_lemma0
  assume i1 %lemma0
  %C_lemma1 = icmp sge i32 %C, 0
  %X_lemma1 = icmp sge i32 %x, 0
  %lemma1 = and i1 %C_lemma1, %X_lemma1
  assume i1 %lemma1
  %cmp = icmp uge i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_uge(i32 %x, i32 %C) {
#0:
  %C_lemma0 = icmp ult i32 %C, 16777216
  %X_lemma0 = icmp ult i32 %x, 16777216
  %lemma0 = or i1 %C_lemma0, %X_lemma0
  assume i1 %lemma0
  %C_lemma1 = icmp sge i32 %C, 0
  %X_lemma1 = icmp sge i32 %x, 0
  %lemma1 = and i1 %C_lemma1, %X_lemma1
  assume i1 %lemma1
  %CFp = sitofp i32 %C to float
  %conv = sitofp i32 %x to float
  %cmp = fcmp oge float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_eq_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp eq i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_eq_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = bitcast i32 %C to float
  %conv = bitcast i32 %x to float
  %cmp = fcmp oeq float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_ne_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp ne i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_ne_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = bitcast i32 %C to float
  %conv = bitcast i32 %x to float
  %cmp = fcmp one float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_slt_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp slt i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_slt_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = bitcast i32 %C to float
  %conv = bitcast i32 %x to float
  %cmp = fcmp olt float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_sgt_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp sgt i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_sgt_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = bitcast i32 %C to float
  %conv = bitcast i32 %x to float
  %cmp = fcmp ogt float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_sle_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp sle i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_sle_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = bitcast i32 %C to float
  %conv = bitcast i32 %x to float
  %cmp = fcmp ole float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_sge_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp sge i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_sge_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = bitcast i32 %C to float
  %conv = bitcast i32 %x to float
  %cmp = fcmp oge float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_ult_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp ult i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_ult_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = bitcast i32 %C to float
  %conv = bitcast i32 %x to float
  %cmp = fcmp olt float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_ule_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp ule i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_ule_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = bitcast i32 %C to float
  %conv = bitcast i32 %x to float
  %cmp = fcmp ole float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_ugt_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp ugt i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_ugt_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = bitcast i32 %C to float
  %conv = bitcast i32 %x to float
  %cmp = fcmp ogt float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!


----------------------------------------
define i1 @src_uge_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %cmp = icmp uge i32 %x, %C
  ret i1 %cmp
}
=>
define i1 @tgt_uge_bitcast(i32 %x, i32 %C) {
#0:
  %C_lemma = icmp ult i32 %C, 2139095041
  %X_lemma = icmp ult i32 %x, 2139095041
  %lemma = and i1 %C_lemma, %X_lemma
  assume i1 %lemma
  %CFp = bitcast i32 %C to float
  %conv = bitcast i32 %x to float
  %cmp = fcmp oge float %conv, %CFp
  ret i1 %cmp
}
Transformation seems to be correct!

Summary:
  20 correct transformations
  0 incorrect transformations
  0 failed-to-prove transformations
  0 Alive2 errors

```

goldsteinn commented on llvm/test/CodeGen/X86/sadd_sat_vec.ll:
```diff
@@ -861,6 +861,9 @@ define <8 x i32> @v8i32(<8 x i32> %x, <8 x i32> %y) nounwind {
 ; AVX1-NEXT:    vinsertf128 $1, %xmm3, %ymm0, %ymm0
+; AVX1-NEXT:    vcvtdq2ps %ymm1, %ymm1
+; AVX1-NEXT:    vpxor %xmm3, %xmm3, %xmm3
+; AVX1-NEXT:    vcmpltps %ymm3, %ymm1, %ymm1
```
Not sure why this is happening. It seems we are determining the compare is dead before but not after.
Not sure what the best way to fix this is.

goldsteinn replied:
Think fixed. Added support for `cmpp` in simplifydemandedbits.

goldsteinn commented on llvm/test/CodeGen/X86/sadd_sat_vec.ll:
```diff
@@ -1059,6 +1062,9 @@ define <16 x i32> @v16i32(<16 x i32> %x, <16 x i32> %y) nounwind {
 ; AVX1-NEXT:    vinsertf128 $1, %xmm5, %ymm0, %ymm0
+; AVX1-NEXT:    vcvtdq2ps %ymm2, %ymm2
+; AVX1-NEXT:    vpxor %xmm5, %xmm5, %xmm5
+; AVX1-NEXT:    vcmpltps %ymm5, %ymm2, %ymm2
```
Same as above

goldsteinn commented on llvm/test/CodeGen/X86/sadd_sat_vec.ll:
```diff
@@ -1067,19 +1073,21 @@ define <16 x i32> @v16i32(<16 x i32> %x, <16 x i32> %y) nounwind {
+; AVX1-NEXT:    vpcmpgtd %xmm7, %xmm1, %xmm1
+; AVX1-NEXT:    vinsertf128 $1, %xmm6, %ymm1, %ymm1
+; AVX1-NEXT:    vcvtdq2ps %ymm3, %ymm3
+; AVX1-NEXT:    vcmpltps %ymm5, %ymm3, %ymm3
```
Same as above.

phoebewang commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -23299,6 +23299,110 @@ static SDValue LowerVSETCC(SDValue Op, const X86Subtarget &Subtarget,
+      } else {
+        // For cvt + signed compare we need  abs(lhs) or abs(rhs) <
+        // MaxConvertableCvt
+        if (OpInRange(1, /*CmpLT*/ true, MaxConvertableCvt) ||
```
`&&`?

goldsteinn replied:
its `||`. 
Proof is something like:
```
define i1 @src_eq(i32 %x, i32 %C) {
  %C_abs = call i32 @llvm.abs.i32(i32 %C, i1 false)
  %X_abs = call i32 @llvm.abs.i32(i32 %x, i1 false)
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  call void @llvm.assume(i1 %lemma)

  %cmp = icmp eq i32 %x, %C
  ret i1 %cmp
}

define i1 @tgt_eq(i32 %x, i32 %C) {
  %C_abs = call i32 @llvm.abs.i32(i32 %C, i1 false)
  %X_abs = call i32 @llvm.abs.i32(i32 %x, i1 false)
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  call void @llvm.assume(i1 %lemma)

  %CFp = sitofp i32 %C to float
  %conv = sitofp i32 %x to float
  %cmp = fcmp oeq float %conv, %CFp
  ret i1 %cmp
}

```

The condition:
```
  %C_abs = call i32 @llvm.abs.i32(i32 %C, i1 false)
  %X_abs = call i32 @llvm.abs.i32(i32 %x, i1 false)
  %C_lemma = icmp ult i32 %C_abs, 16777216
  %X_lemma = icmp ult i32 %X_abs, 16777216
  %lemma = or i1 %C_lemma, %X_lemma
  call void @llvm.assume(i1 %lemma)

```
is all `||`.

phoebewang replied:
Isn't a `abs(x) < c` is `c < x < -c`? Otherwise `x` can be any possible value for a `x < c || x > -c`.

goldsteinn replied:
note that the `x > -c` is `x u> -c` which is a pretty tight bound.

phoebewang replied:
Oh, I see. Comparing signed value with unsigned operator. It's a bit difficult for me to notice that :)

phoebewang commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -23299,6 +23299,110 @@ static SDValue LowerVSETCC(SDValue Op, const X86Subtarget &Subtarget,
+        // MaxConvertableCvt
+        if (OpInRange(1, /*CmpLT*/ true, MaxConvertableCvt) ||
+            OpInRange(1, /*CmpLT*/ false, -MaxConvertableCvt) ||
+            OpInRange(0, /*CmpLT*/ true, MaxConvertableCvt) ||
```
ditto.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -23299,6 +23299,110 @@ static SDValue LowerVSETCC(SDValue Op, const X86Subtarget &Subtarget,
 
+  // We get bad codegen for v8i32 compares on avx targets (without avx2) so if
+  // possible convert to a v8f32 compare.
+  if (VTOp0.getVectorElementType() == MVT::i32 && VTOp0 == MVT::v8i32 &&
```
Just test VTOp0 == MVT::v8i32 for ? VTOp0.getVectorElementType() == MVT::i32 looks unnecessary.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -23299,6 +23299,110 @@ static SDValue LowerVSETCC(SDValue Op, const X86Subtarget &Subtarget,
+    bool OkayCvt = false;
+    bool OkayBitcast = false;
+
+    // For cvt up to 1 << (Significand Precision)
```
Would it make sense to use the APFloat semantics instead of hard coded numbers here?

goldsteinn replied:
Not 100% sure, I reached the constants in the proofs via trial and error then realized these numbers are the most logical bound. Not certain though that they are indeed generalizable.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -23299,6 +23299,110 @@ static SDValue LowerVSETCC(SDValue Op, const X86Subtarget &Subtarget,
+    auto OpInRange = [&DAG, Op, &KnownOps](unsigned OpNo, bool CmpLT,
+                                           const APInt Bound) {
+      if (!KnownOps[OpNo].has_value())
+        KnownOps[OpNo] = DAG.computeKnownBits(Op.getOperand(OpNo));
```
Why do you want to cache KnownOps[] - shouldn't we just call both at the start and be done with it?

goldsteinn replied:
We can get away with only doing `Op1` if its a small constants (which is the most likely path).

RKSimon reviewed (commented):
I'm not sure about the OkayBitcast codepath - I was just expecting some simple knownbits min/max value tests and SINT_TO_FP calls that would also handle constant folding.

phoebewang commented:
I recall I read someone commented with concern when DAZ is set, fcmp will always return equal for denormal values, but I cannot find the comment now. Is this a real concern? Did we fix it?

goldsteinn commented:
> > I recall I read someone commented with concern when DAZ is set, fcmp will always return equal for denormal values, but I cannot find the comment now. Is this a real concern? Did we fix it?
> 
> Looks like that would be the result of DAZ, but didn't test: https://stackoverflow.com/questions/54047415/do-denormal-flags-like-denormals-are-zero-daz-affect-comparisons-for-equality
> 
> Ill update with `sub` flag.

Actually, I don't think this is an issue since we don't handle `eq`/`ne`

goldsteinn commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -2471,6 +2474,8 @@ bool TargetLowering::SimplifyDemandedBits(
     if (SimplifyDemandedBits(Src, InDemandedBits, InDemandedElts, Known, TLO,
                              Depth + 1))
       return true;
+
```
Yeah, thank you!

phoebewang commented:
> > > I recall I read someone commented with concern when DAZ is set, fcmp will always return equal for denormal values, but I cannot find the comment now. Is this a real concern? Did we fix it?
> > 
> > 
> > Looks like that would be the result of DAZ, but didn't test: https://stackoverflow.com/questions/54047415/do-denormal-flags-like-denormals-are-zero-daz-affect-comparisons-for-equality
> > Ill update with `sub` flag.
> 
> Actually, I don't think this is an issue since we don't handle `eq`/`ne`

It affects all predicates, not only `eq`/`ne`, see https://godbolt.org/z/jxs5s5qde

goldsteinn commented:
> > > > I recall I read someone commented with concern when DAZ is set, fcmp will always return equal for denormal values, but I cannot find the comment now. Is this a real concern? Did we fix it?
> > > 
> > > 
> > > Looks like that would be the result of DAZ, but didn't test: https://stackoverflow.com/questions/54047415/do-denormal-flags-like-denormals-are-zero-daz-affect-comparisons-for-equality
> > > Ill update with `sub` flag.
> > 
> > 
> > Actually, I don't think this is an issue since we don't handle `eq`/`ne`
> 
> It affects all predicates, not only `eq`/`ne`, see https://godbolt.org/z/jxs5s5qde

ah.

We don't naturally have `subnormal` fpclass in DAG, so think we just need to disable the simplifications involving `Op1C == 0` if `-mdaz-ftz`? Or would we need to disable it regardless unless we can prove `Op0` is normal?

anematode commented:
We should probably disable it regardless of the compiler settings because it'd be weird for FP flags to invalidate what looks like integer comparisons.... In my comment where I brought this up I posted a way to compare f32 for equality/inequality regardless of range, but it fails to generalize. I think your converting-to-float method is totally fine though.

goldsteinn commented:
> We should probably disable it regardless of the compiler settings because it'd be weird for FP flags to invalidate what looks like integer comparisons.... In my comment where I brought this up I posted a way to compare i32 for equality/inequality regardless of range, but it fails to generalize. I think your converting-to-float method is totally fine though.

What do you mean by invalidate? The issue is simplifying a float signbit comparison to just return the float. DAZ can make the simplification incorrect.

anematode commented:
Maybe we're talking about different things then, sorry. I thought this was in relation to when we can or cannot bitcast integers to floats and do comparison as floats. For example, I think the restriction on C in `src_ule_bitcast -> tgt_ule_bitcast` is that it must be between 1 << 23 and the existing bound, not just 0 and the bound.

goldsteinn commented:
> Maybe we're talking about different things then, sorry. I thought this was in relation to when we can or cannot bitcast integers to floats and do comparison as floats. For example, I think the restriction on C in `src_ule_bitcast -> tgt_ule_bitcast` is that it must be between 1 << 23 and the existing bound, not just 0 and the bound.

Ah that is another issue...

So there are two issues with DAZ:
1) `bitcast` only check ops < inf_bits
2)  `simplifySExtOfDecomposedSetCC` will simplify op `lt`/`gt` 0.0 (what I thought you where referencing).

Think both probably need to be fixed.

Still re:
> We should probably disable it regardless of the compiler settings because it'd be weird for FP flags to invalidate what looks like integer comparisons....

Do you mean drop the optimization (if we can't prove normal)? Don't really see whats the issue with seeing if we are compiling with DAZ.

anematode commented:
Yeah, but only the bitcast optimization and only if we can't prove the operands are normal. I never looked at `simplifySExtOfDecomposedSetCC` – sorry for the confusion. The issue is that in general I don't think we can rely on DAZ not being set, just because we didn't compile with it enabled. For example, if you link a library that's compiled with `-ffast-math` it sometimes will enable DAZ without you noticing. (The curse of global state!) Then it would be doubly unexpected to have what looks like integer math give spurious results.

goldsteinn commented:
> @goldsteinn Would you be willing to split the bitcast comparisons into a followup patch and just get the int2fp version done first?

Will do.

goldsteinn commented:
> > @goldsteinn Would you be willing to split the bitcast comparisons into a followup patch and just get the int2fp version done first?
> 
> Will do.

Err, I think the fix for both is pretty trivial. I'll post 1 more time with complete. If any more concerns
ill drop the bitcast aspect.

goldsteinn commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -2313,6 +2313,26 @@ bool TargetLowering::SimplifyDemandedBits(
+
+    Known = TLO.DAG.computeKnownBits(Op, DemandedElts, Depth);
+    break;
+  }
```
This is necessary to clean up regression of extra `cvt` instructions left behind if we simplify the transformation.

goldsteinn replied:
added explicit tests.

RKSimon replied:
Cheers - any luck with pulling this out entirely?

goldsteinn replied:
Ah, done, see #85445

goldsteinn commented on llvm/test/CodeGen/X86/setcc-lowering.ll:
```diff
@@ -9,10 +9,11 @@
+; AVX-NEXT:    vandps {{\.?LCPI[0-9]+_[0-9]+}}(%rip), %ymm0, %ymm0
+; AVX-NEXT:    vcvtdq2ps %ymm0, %ymm0
+; AVX-NEXT:    vxorps %xmm1, %xmm1, %xmm1
+; AVX-NEXT:    vcmpeqps %ymm1, %ymm0, %ymm0
```
This is the one remaining regression pattern. Basically if demandedelts is only contained the xmm, this will be a regression.

I was thinking maybe move the entire transform to `SimplifyDemandedBits` and only do if `DemandedElts` contains `ymm`, but the transform is pretty x86_64 exclusive so wasn't sure if it fit.

Any ideas @RKSimon

RKSimon replied:
I'll take a look

RKSimon replied:
I'm not sure how we can easily avoid the conversion, but I've pushed a couple of changes that should help - please can you rebase? I think the vcvtdq2ps might still be a ymm but we should be getting better.

goldsteinn replied:
Rebased and at least it stays in `xmm` now. I think what we could do, is I could make the transform reversible. Then we could have rules for int->float and float->int.

goldsteinn replied:
@RKSimon , what do you think?

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -2313,6 +2313,26 @@ bool TargetLowering::SimplifyDemandedBits(
+    EVT InnerVT = Op.getOperand(0).getValueType();
+    const APInt &DemandedBitsToUse =
+        TLO.DAG.getTargetLoweringInfo().isBitcastFree(VT, InnerVT)
+            ? OriginalDemandedBits
```
You can't safely use OriginalDemandedBits as it won't account for multiple uses - you will need to move this inside SimplifyMultipleUseDemandedBits

goldsteinn replied:
Fixed.

RKSimon commented on llvm/test/CodeGen/X86/setcc-lowering.ll:
```diff
@@ -10,10 +10,11 @@
-; AVX1-NEXT:    vpxor %xmm1, %xmm1, %xmm1
-; AVX1-NEXT:    vpcmpeqd %xmm1, %xmm0, %xmm0
+; AVX1-NEXT:    vxorps %xmm1, %xmm1, %xmm1
+; AVX1-NEXT:    vcmpeqps %xmm1, %xmm0, %xmm0
```
Still not perfect, but I'm not sure how much more we should try to do.

Maybe worth folding `extract_subvector(sint_to_fp())` -> `sint_to_fp(extract_subvector())`, but not sure if we should bother folding the vcmpeqps back to vpcmpeq again....

goldsteinn replied:
Made a patch for that. It has a lot of changes so I'll post if independently.

goldsteinn replied:
Going to drop actually, doesn't seem to yield in benefits for any targets.

anematode commented:
Reposting here. If the input bound checks fail, `vpcmpeqd` can be emulated with no restrictions on the input bounds:

```
vpxor xmm0, a, b
vcvtdq2ps xmm1, xmm0
vcmpps xmm1, zero, 0
```

goldsteinn commented:
> Reposting here. If the input bound checks fail, `vpcmpeqd` can be emulated with no restrictions on the input bounds:
> 
> ```
> vpcmpeqd result, a, b
> =>
> vpxor ymm0, a, b
> vcvtdq2ps ymm1, ymm0
> vcmpps result, ymm1, zero, 0
> ```

Ah nice :)

Ill add todo for that.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -23296,6 +23303,136 @@ static SDValue LowerVSETCC(SDValue Op, const X86Subtarget &Subtarget,
+
+    // For cvt up to 1 << (Significand Precision), (1 << 24 for ieee float)
+    const APInt MaxConvertableCvt =
+        APInt(32, (1U << APFloat::semanticsPrecision(Sem)));
```
APint::getOneBitSet(32, APFloat::semanticsPrecision(Sem))

RKSimon commented on llvm/test/CodeGen/X86/vector-reduce-fmaximum.ll:
```diff
@@ -816,64 +816,6 @@ define float @test_v16f32(<16 x float> %a0) {
 ; SSE41-NEXT:    orps %xmm4, %xmm0
 ; SSE41-NEXT:    retq
 ;
-; AVX-LABEL: test_v16f32:
```
Looks like its missing a non-common check-prefix

goldsteinn replied:
regening, not sure what happened.

RKSimon replied:
still not fixed

RKSimon commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -625,6 +625,8 @@ class TargetLoweringBase {
     return isLoadBitCastBeneficial(StoreVT, BitcastVT, DAG, MMO);
   }
 
+  virtual bool isBitcastFree(EVT DstVT, EVT SrcVT) const { return false; }
```
Is this still necessary?

goldsteinn replied:
Its for multiuse case, since we are creating an additional instruction need it be free no?.

RKSimon replied:
Do you get instruction bloat without it? Vector bitcasts are typically free on most targets.

goldsteinn replied:
I didn't notice any issue, the check just seemed basic due-diligence. I can make the default "true" if you think that makes more sense.

RKSimon replied:
I'd drop it entirely - we already have plenty of cases where we just bitcast between types.

RKSimon commented:
rebase?

goldsteinn commented:
> rebase?

Done.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -30,6 +30,7 @@
 #include "llvm/Analysis/ObjCARCUtil.h"
 #include "llvm/Analysis/ProfileSummaryInfo.h"
 #include "llvm/Analysis/VectorUtils.h"
+#include "llvm/CodeGen/ISDOpcodes.h"
```
This shouldn't be necessary

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -41073,6 +41204,150 @@ static SDValue combineShuffle(SDNode *N, SelectionDAG &DAG,
+  } else if (auto *C = dyn_cast<ConstantSDNode>(Op1)) {
+    Op1C = C->getAPIntValue();
+  } else if (ISD::isConstantSplatVector(Op1.getNode(), Op1C)) {
+    // Pass
```
Pass? Make it clear that isConstantSplatVector sets Op1C

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -42442,13 +42730,24 @@ SDValue X86TargetLowering::SimplifyMultipleUseDemandedBitsForTargetNode(
+            DAG, DL, ISD::SETGT, Op.getOperand(0), Op.getOperand(1),
+            DemandedBits, /*AllowNOT*/ false, Depth))
+      return R;
+    break;
```
Why is this case necessary? I thought it was just to convert a XMM X86ISD::CMPP back to an integer?

goldsteinn replied:
Not, its a common helper to do the simplifications for float/int.

I could add float->int, however.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -23353,6 +23353,136 @@ static SDValue LowerVSETCC(SDValue Op, const X86Subtarget &Subtarget,
+        Res = KnownBits::ult(*KnownOps[OpNo], KnownBits::makeConstant(Bound));
+      else
+        Res = KnownBits::ugt(*KnownOps[OpNo], KnownBits::makeConstant(Bound));
+      return Res.has_value() && *Res;
```
Maybe `return Res.value_or(false)`?

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -41125,6 +41255,154 @@ static SDValue combineShuffle(SDNode *N, SelectionDAG &DAG,
 
+// Simplify a decomposed (sext (setcc)). Assumes prior check that
+// bitwidth(sext)==bitwidth(setcc operands).
+static SDValue simplifySExtOfDecomposedSetCCImpl(
```
Any hope of pulling this out into its own patch?

goldsteinn replied:
Will do

goldsteinn replied:
@RKSimon seperate commit in this PR, or new PR as a whole?

goldsteinn replied:
Can do, the only issue is that basically its functionality is dependent on this pr. Otherwise we never really get the necessary floating point flags to assist with the simplifcation.

goldsteinn replied:
So this PR will have some regressions in it, ill post the other PR as rebased ontop of this one.

goldsteinn replied:
Okay, dropped the simplifcations from this patch. See: https://github.com/llvm/llvm-project/pull/84360 (based on this one).

goldsteinn replied:
Note, there are some regressions in this patch now but they should be cleared up in #84360.

Also note, the reason #84360 is second is so that we can properly test the `cmpp` code. Without this patch we never get the necessary FMF flags to actually do the `cmpp` simplifications. If you think that is unnecessary I can swap the order (#84360 does stand on its own mostly for is simplifications on `pcmpgt`).

goldsteinn committed:
[X86] Try Folding `icmp` of `v8i32` -> `fcmp` of `v8f32` on AVX

Fixes: #82242

The idea is that AVX doesn't support comparisons for `v8i32` so it
splits the comparison into 2x `v4i32` comparisons + reconstruction of
the `v8i32`.

By converting to a float, we can handle the comparison with 1/2
instructions (1 if we can `bitcast`, 2 if we need to cast with
`sitofp`).

The Proofs: https://alive2.llvm.org/ce/z/AJDdQ8
Timeout, but they can be reproduced locally.

goldsteinn committed:
[X86] Improve helper for simplifying demanded bits of compares

We currently only handle a single case for `pcmpgt`. This patch
extends that to work for `cmpp` and handles comparitors more
generically.

RKSimon reviewed (commented):
Any luck with the remaining regressions?

goldsteinn commented:
> Any luck with the remaining regressions?

Most of the regressions are fixed with #84360, the only outstanding one
is when we later simplify to `xmm` w/ demandedelts in which case we
have converted from int -> float without any benefit. For example:
https://github.com/llvm/llvm-project/pull/84360/files#diff-ec93308ebc948e632bea97e65a5aeedc4342aab5fa49d2c4a43be983131fba47R17

Thats tough to get around, as float -> int is much more difficult to
pull off.

goldsteinn commented:
> > > Any luck with the remaining regressions?
> > 
> > 
> > Most of the regressions are fixed with #84360, the only outstanding one is when we later simplify to `xmm` w/ demandedelts in which case we have converted from int -> float without any benefit. For example: https://github.com/llvm/llvm-project/pull/84360/files#diff-ec93308ebc948e632bea97e65a5aeedc4342aab5fa49d2c4a43be983131fba47R17
> > Thats tough to get around, as float -> int is much more difficult to pull off.
> 
> Im trying to refactor so the code arbitrarily works to cvt int <-> fp compares. Then we can undo as needed.

That doesn't really help. Maybe this isn't worth doing after all?

goldsteinn commented:
@rksimon, thoughts on dropping this? Do you think its worth the complexity?

RKSimon commented:
> @RKSimon, thoughts on dropping this? Do you think its worth the complexity?

Do you know which part of the fold is causing the most regressions? #82242 was for the limited case of when we've masked the lower bits for comparison - if we just handled that, what does it look like then?

goldsteinn commented:
> > @RKSimon, thoughts on dropping this? Do you think its worth the complexity?
> 
> Do you know which part of the fold is causing the most regressions? #82242 was for the limited case of when we've masked the lower bits for comparison - if we just handled that, what does it look like then?

Most of the regressions are caused by not simplifying `CMPP` which #84360 cleans up.
The remainder of when we simplify `ymm` -> `xmm` after the transform which is harder
to undo.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -23391,6 +23391,136 @@ static SDValue LowerVSETCC(SDValue Op, const X86Subtarget &Subtarget,
+        // or rhs < MaxConvertableCvt
+
+        if (OpInRange(1, /*CmpLT*/ true, APInt::getSignedMinValue(32)) &&
+            OpInRange(0, /*CmpLT*/ true, APInt::getSignedMinValue(32)) &&
```
@goldsteinn The comment says both lhs/rhs >= 0 but the comparisons are with MIN_INT ?

goldsteinn replied:
its unsigned compare. So `foo u< MIN_INT` -> `foo s>= 0`.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -23391,6 +23391,136 @@ static SDValue LowerVSETCC(SDValue Op, const X86Subtarget &Subtarget,
+            OpInRange(1, /*CmpLT*/ false, -MaxConvertableCvt) ||
+            OpInRange(0, /*CmpLT*/ true, MaxConvertableCvt) ||
+            OpInRange(0, /*CmpLT*/ false, -MaxConvertableCvt))
+          OkayCvt = true;
```
Don't we need signed comparisons here - OpInRange just does unsigned comparisons? Ideally we'd use NumSignBits checks as an alternative to also check for being in signed bounds.

RKSimon commented:
It'd be good to get #82242 fixed - have you any thoughts on how we can reduce the scope of this patch to avoid having to chase regressions so much?

goldsteinn commented:
> It'd be good to get #82242 fixed - have you any thoughts on how we can reduce the scope of this patch to avoid having to chase regressions so much?

I think most of the scope ends up being there too fix regressions. The fundemental issue is that going backwards is a lot more difficult and we are able to do some simplifications on the integer versions that we can't on the float ones.

Not really sure how to address them TBH.

RKSimon commented:
@goldsteinn close this? I'm not sure how much more we can push the AVX1 256-bit icmp support
